File: R/gh_stats_card.R


# Stats Card --------------------------------------------------------------

#' Create GitHub Stats Card 
#'
#' @param username Your Github username
#' @param hide If not `NULL`, provide stats to hide, one of: stars, commits, prs, issues, contribs.
#' @param count_private (Boolean) `TRUE` to add the count of all your private contributions to the total commits count
#' @param show_icons (Boolean) `TRUE` to enable icons
#' @param theme If not `NULL`, provide build in theme names, e.g., one of: dark, radical, merko, gruvbox, tokyonight, onedark, cobalt, synthwave, highcontrast, dracula.
#'
#' @return A character rendering GitHub Stats Card to be inserted into R Markdown "github_document"
#' @export
#' @source https://github.com/anuraghazra/github-readme-stats
#' 
gh_stats_card <- function(username,
                          hide = NULL,
                          count_private = FALSE,
                          show_icons = FALSE,
                          theme = NULL
){
  
  url <- gh_stats_card_url(username = username, 
                           hide = hide, 
                           count_private = count_private, 
                           show_icons = show_icons, 
                           theme = theme)
  
  ### Build Markdown Link
  glue::glue("[![{username} GitHub stats]({url})](https://github.com/{username}/{username})")
  
}


# Github Stats Card URL ---------------------------------------------------



#' Create GitHub Stats Card URL 
#'
#' @param username Your Github username
#' @param hide If not `NULL`, provide stats to hide, one of: stars, commits, prs, issues, contribs.
#' @param count_private (Boolean) `TRUE` to add the count of all your private contributions to the total commits count
#' @param show_icons (Boolean) `TRUE` to enable icons
#' @param theme If not `NULL`, provide build in theme names, e.g., one of: dark, radical, merko, gruvbox, tokyonight, onedark, cobalt, synthwave, highcontrast, dracula.
#'
#' @return A URL of GitHub Stats Card
#' @export
#' @source https://github.com/anuraghazra/github-readme-stats
#' 
gh_stats_card_url <- function(username,
                              hide = NULL,
                              count_private = FALSE,
                              show_icons = FALSE,
                              theme = NULL
) {
  
  ## Hide Individual Stats
  hide <- if (is.null(hide)) {
    ""
  } else {
    ## Options: Stats to Hide
    stats_nm <- match.arg(hide,
                          choices = c("stars", "commits", "prs", "issues", "contribs")
    )
    paste0("&hide=", paste(stats_nm, collapse = ","))
  }
  
  ## Count Private
  count_private <- if (count_private) {
    "&count_private=true"
  } else {
    ""
  }
  
  ## Show Icons
  show_icons <- if (show_icons) {
    "&show_icons=true"
  } else {
    ""
  }
  ## Theme
  theme <- if (is.null(theme)) {
    ""
  } else {
    paste0("&theme=", theme)
  }
  
  ### Build Markdown Link
  glue::glue("https://github-readme-stats.vercel.app/api?username={username}{hide}{count_private}{show_icons}{theme}")
}

File: src/gh_stats_card.py
"""GitHub Stats Card URL generation."""


def gh_stats_card_url(
    username: str,
    hide: list[str] | None = None,
    count_private: bool = False,
    show_icons: bool = False,
    theme: str | None = None,
) -> str:
    """
    Create GitHub Stats Card URL.

    Parameters
    ----------
    username : str
        Your GitHub username
    hide : list[str] | None, optional
        Stats to hide. Options: "stars", "commits", "prs", "issues", "contribs"
        by default None
    count_private : bool, optional
        Add count of all private contributions to total commits count,
        by default False
    show_icons : bool, optional
        Enable icons on the card, by default False
    theme : str | None, optional
        Built-in theme name (e.g., "dark", "radical", "merko", "gruvbox",
        "tokyonight", "onedark", "cobalt", "synthwave", "highcontrast", "dracula"),
        by default None

    Returns
    -------
    str
        URL of GitHub Stats Card

    Notes
    -----
    Based on: https://github.com/anuraghazra/github-readme-stats

    Examples
    --------
    >>> gh_stats_card_url("octocat")
    'https://github-readme-stats.vercel.app/api?username=octocat'

    >>> gh_stats_card_url("octocat", count_private=True, show_icons=True, theme="radical")
    'https://github-readme-stats.vercel.app/api?username=octocat&count_private=true&show_icons=true&theme=radical'
    """
    # Validate hide options
    if hide is not None:
        valid_hide_options = {"stars", "commits", "prs", "issues", "contribs"}
        invalid_options = set(hide) - valid_hide_options
        if invalid_options:
            raise ValueError(
                f"Invalid hide options: {invalid_options}. "
                f"Valid options are: {valid_hide_options}"
            )

    # Build query parameters
    params = []

    if hide:
        hide_str = ",".join(hide)
        params.append(f"hide={hide_str}")

    if count_private:
        params.append("count_private=true")

    if show_icons:
        params.append("show_icons=true")

    if theme:
        params.append(f"theme={theme}")

    # Construct URL
    base_url = f"https://github-readme-stats.vercel.app/api?username={username}"
    if params:
        query_string = "&".join(params)
        return f"{base_url}&{query_string}"

    return base_url


def gh_stats_card(
    username: str,
    hide: list[str] | None = None,
    count_private: bool = False,
    show_icons: bool = False,
    theme: str | None = None,
) -> str:
    """
    Create GitHub Stats Card markdown link.

    Parameters
    ----------
    username : str
        Your GitHub username
    hide : list[str] | None, optional
        Stats to hide, by default None
    count_private : bool, optional
        Add count of private contributions, by default False
    show_icons : bool, optional
        Enable icons on the card, by default False
    theme : str | None, optional
        Built-in theme name, by default None

    Returns
    -------
    str
        Markdown link rendering GitHub Stats Card

    Notes
    -----
    Based on: https://github.com/anuraghazra/github-readme-stats
    """
    url = gh_stats_card_url(
        username=username,
        hide=hide,
        count_private=count_private,
        show_icons=show_icons,
        theme=theme,
    )

    return f"[![{username} GitHub stats]({url})](https://github.com/{username}/{username})"
